File: server/typescript/packages/metadata/src/registry-manifest.ts
// SP-G Registry Conformance — the TS reference emitter.
//
// Walks an assembled core TypeRegistry and serializes the LOGICAL metamodel
// vocabulary as a canonical, fully-sorted, byte-stable JSON manifest. This is
// the single-source contract the other four ports (C#, Java, Kotlin, Python)
// must byte-match — a structural gate against the SP-C class of silent
// vocabulary drift (a port's registry diverging — wrong attr names, missing
// subtypes, different required-ness — with every behavioral corpus still green).
//
// The IN/OUT boundary is documented in fixtures/registry-conformance/README.md.
// In short, the manifest emits: type.subType + per-type/per-attr `description`
// (FR-033) + attrs[{name, valueType, isArray, required, allowedValues?,
// description, rules?, example?, whenToUse?}] + the structural constraint graph
// (children / parents / cardinality, FR-033) + commonAttrs + defaultSubTypes.
// `allowedValues` (ADR-0036 Wave 1, decision 5) is the closed value-set of a
// closed-enum attr — emitted ONLY when the attr declares one, OMITTED for open /
// format-validated attrs (@currency, @locale). It byte-gates closed vocabularies
// cross-port (the gate that catches a value-set silently drifting between ports).
// EXCLUDED (per-port-physical or not-universally-tracked-on-the-registry):
// factories/native bindings; AttrSchema.default.
//
// FR-033 Task 5 GREW the manifest: it now also emits the documentation surface
// (every type/attr carries a required, non-empty `description`; optional `rules`/
// `example`/`whenToUse`) AND the full structural constraint graph (each type's
// `children` from childRules — childType/childSubType/childName + optional
// cardinality min/max/named — and optional `parents`). Growing the canonical
// RED-flags the other four ports' registry-conformance until they reconcile
// (the intended intermediate state, same as the FR-032 sweep).

import type { AttrSchema, ChildRule, TypeDefinition, TypeRegistry } from "./registry.js";
import { ATTR_SUBTYPE_STRING, ATTR_SUBTYPE_STRINGARRAY } from "./core/attr/attr-constants.js";
import {
  EXCLUDED_PER_TYPE_ATTRS,
  ExclusionReason,
  isExcludedTypeSubType,
  manifestRequiredOverride,
} from "./registry-manifest-exclusions.js";

/** One attribute in the manifest — the logical, cross-port-identical facet. */
interface ManifestAttr {
  name: string;
  /** The attr's SCALAR value-type subtype, or null for a polymorphic/untyped attr (e.g. @default). */
  valueType: string | null;
  /** True for an array-valued attr (a list of the scalar `valueType`); the orthogonal array axis. */
  isArray: boolean;
  required: boolean;
  /**
   * The closed value-set of a closed-enum attr (ADR-0036 Wave 1, decision 5).
   * Emitted ONLY when the attr declares a non-empty `allowedValues` set; OMITTED
   * for open / format-validated attrs (e.g. @currency ISO-4217, @locale BCP-47).
   * This byte-gates closed-enum vocabularies cross-port — the gate that catches a
   * value-set drifting between ports (the failure mode the dbColumnType slim hit).
   */
  allowedValues?: readonly string[];
  /** FR-033 — human/AI-facing description of the attribute (required, non-empty). */
  description: string;
  /** FR-033 — prose documenting the complex rules enforced in code. Emitted only when present. */
  rules?: string;
  /** FR-033 — an example value. Emitted only when present. */
  example?: string;
  /** FR-033 — guidance on when to reach for this attribute. Emitted only when present. */
  whenToUse?: string;
}

/** One structural child rule of a type (FR-033 constraint graph). */
interface ManifestChild {
  /** The admitted child `type` (`"*"` = any). */
  childType: string;
  /** The admitted child subType — a single subtype, `"*"` (any), or a list of admitted subtypes. */
  childSubType: string | readonly string[];
  /** The admitted child name (`"*"` = any). */
  childName: string;
  /** Cardinality lower bound — emitted only when defined on the rule. */
  min?: number;
  /** Cardinality upper bound (`null` = unbounded) — emitted only when defined on the rule. */
  max?: number | null;
  /** Whether the child must carry an explicit name — emitted only when defined on the rule. */
  named?: boolean;
}

/** One registered (type, subType) in the manifest, with its docs + attrs + constraint graph. */
interface ManifestType {
  type: string;
  subType: string;
  /** FR-033 — human/AI-facing description of the type/subType (required, non-empty). */
  description: string;
  /** FR-033 — prose documenting the complex rules enforced in code. Emitted only when present. */
  rules?: string;
  /** FR-033 — an example. Emitted only when present. */
  example?: string;
  /** FR-033 — guidance on when to reach for this type/subType. Emitted only when present. */
  whenToUse?: string;
  attrs: ManifestAttr[];
  /** FR-033 — the type's structural child rules (sorted), the constraint graph. */
  children: ManifestChild[];
  /** FR-033 — the child-side placement claim (sorted). Emitted only when present + non-empty. */
  parents?: string[];
}

/**
 * The Metamodel spec version — the shared, rolled-up name for the frozen core
 * vocabulary (ADR-0035 §2 / docs/1.0-readiness.md C4). It is a version TAG on this
 * manifest (the byte-exact bill of materials), NOT a per-provider or per-file marker:
 * every port emits the same string, asserted by registry-conformance. Pre-1.0 the
 * vocabulary is still `0.x` (semver = unstable, may change) but largely settled going
 * into the quiet period; the 1.0 cut (readiness G1) freezes it to `"1.0"`. Decoupled
 * from every package line — it is the spec version, not a package version.
 *
 * The current value is the declaration below and nothing else. Do not restate it in
 * prose: `scripts/check-metamodel-version.mjs --set` rewrites the declaration, so any
 * copy in a comment goes stale on the very next bump (it said `"0.9"` here while the
 * constant read `"0.10"`). Bump with that script — never by hand — so the manifest and
 * all four port constants move together.
 */
export const METAMODEL_VERSION = "0.13";

/** The full canonical manifest. All collections are sorted for byte-stability. */
interface RegistryManifest {
  /** The Metamodel spec version (first key — a header). See {@link METAMODEL_VERSION}. */
  metamodelVersion: string;
  types: ManifestType[];
  commonAttrs: ManifestAttr[];
  defaultSubTypes: Record<string, string>;
}

/** ASCII-string compare so the sort is locale-independent and byte-stable across ports. */
function compareStrings(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/**
 * Normalize one AttrSchema to the manifest's logical attr shape — decomposing
 * array-ness into a scalar `valueType` + an orthogonal `isArray` flag.
 *
 * An attr is array-valued when its schema sets `isArray: true` (the cross-port
 * model). For resilience, a legacy `valueType: "stringarray"` token is also
 * decomposed to `{ valueType: "string", isArray: true }` — so no `stringarray`
 * token ever reaches the manifest. A polymorphic attr (no valueType) is `null` +
 * non-array.
 */
function toManifestAttr(attr: AttrSchema): ManifestAttr {
  const isLegacyStringArray = attr.valueType === ATTR_SUBTYPE_STRINGARRAY;
  const isArray = attr.isArray === true || isLegacyStringArray;
  // The scalar value-type: a legacy stringarray token collapses to "string";
  // otherwise the declared valueType (omitted → null for polymorphic attrs).
  const valueType = isLegacyStringArray
    ? ATTR_SUBTYPE_STRING
    : (attr.valueType ?? null);
  // FR-033: the documentation surface (`description` required + non-empty;
  // `rules`/`example`/`whenToUse` emitted ONLY when present) follows the
  // existing facets, preserving key order for byte-stability.
  // The closed value-set, emitted only when present + non-empty (decision 5).
  // AttrValue is string | number | boolean | array; a closed enum is always a set
  // of scalar string members, so stringify each for a stable cross-port surface.
  // Spread BEFORE `description` so the JSON key order is name → valueType →
  // isArray → required → (allowedValues?) → description → (rules?/example?/whenToUse?).
  const hasAllowed = attr.allowedValues !== undefined && attr.allowedValues.length > 0;
  const out: ManifestAttr = {
    name: attr.name,
    valueType,
    isArray,
    required: attr.required,
    ...(hasAllowed ? { allowedValues: attr.allowedValues!.map((v) => String(v)) } : {}),
    description: attr.description,
  };
  if (attr.rules !== undefined) out.rules = attr.rules;
  if (attr.example !== undefined) out.example = attr.example;
  if (attr.whenToUse !== undefined) out.whenToUse = attr.whenToUse;
  return out;
}

/** The canonical sort key for a child rule: childType, then the childSubType
 *  string (or comma-joined list), then childName — ASCII codepoint compare. */
function childSubTypeKey(childSubType: string | readonly string[]): string {
  return Array.isArray(childSubType) ? childSubType.join(",") : (childSubType as string);
}

/**
 * Normalize one structural ChildRule to the manifest's child shape (FR-033).
 * Cardinality (`min`/`max`/`named`) is emitted ONLY when defined on the rule —
 * legacy wildcard rules leave them undefined and must NOT fabricate cardinality.
 * `max` may legitimately be `null` (unbounded); `null` is emitted, `undefined`
 * is omitted.
 */
function toManifestChild(rule: ChildRule): ManifestChild {
  const out: ManifestChild = {
    childType: rule.childType,
    childSubType: rule.childSubType,
    childName: rule.childName,
  };
  if (rule.min !== undefined) out.min = rule.min;
  if (rule.max !== undefined) out.max = rule.max;
  if (rule.named !== undefined) out.named = rule.named;
  return out;
}

/** Sort the constraint graph by (childType, childSubTypeKey, childName) — ASCII. */
function sortedChildren(rules: readonly ChildRule[]): ManifestChild[] {
  return rules
    .map(toManifestChild)
    .sort(
      (a, b) =>
        compareStrings(a.childType, b.childType) ||
        compareStrings(childSubTypeKey(a.childSubType), childSubTypeKey(b.childSubType)) ||
        compareStrings(a.childName, b.childName),
    );
}

/** Sort attrs by name (ascending, ASCII). */
function sortedAttrs(attrs: readonly AttrSchema[]): ManifestAttr[] {
  return attrs
    .map(toManifestAttr)
    .sort((a, b) => compareStrings(a.name, b.name));
}

/**
 * Sort PER-TYPE attrs, keeping only those the explicit classification marks
 * INCLUDED (logical cross-port vocabulary). A per-type attr classified with any
 * `ExclusionReason` (structural keyword, native binding, per-type `description`
 * commonAttr dup) is carved out — for a documented reason, never a silent
 * name-match. The filter is a no-op for TS (which never registers the carved-out
 * names as per-type attrs); it is applied uniformly so the cross-port contract
 * is explicit. NOTE: `description` is filtered ONLY here — it remains in the
 * `commonAttrs` block (built via `sortedAttrs`, unfiltered).
 */
function sortedPerTypeAttrs(
  attrs: readonly AttrSchema[],
  type: string,
  subType: string,
): ManifestAttr[] {
  return sortedAttrs(attrs.filter((a) => classifyPerTypeAttr(a.name) === INCLUDED)).map(
    (attr) => {
      // FR-024-pending requiredness override (the attr-level analogue of the
      // Fr024Pending row carve-out): the TS registry already registers the
      // FR-024 requiredness, but the manifest keeps emitting the pre-FR-024
      // agreed value until the Phase-E atomic all-ports flip.
      const required = manifestRequiredOverride(type, subType, attr.name);
      return required === undefined ? attr : { ...attr, required };
    },
  );
}

/**
 * The boundary classifier (Wave 3b). For every per-type attr the emitter
 * encounters, the in/out decision is an EXPLICIT classification — never a silent
 * default. Returns either an `ExclusionReason` (carved out, with a documented
 * category) or `INCLUDED` (logical cross-port vocabulary). It is TOTAL: there is
 * no "unclassified" third state, so a port can never silently let an unreasoned
 * facet through — the self-documenting property the conformance test asserts.
 *
 * This replaces the prior tautology — a bare `EXCLUDED.has(name)` whose negative
 * branch silently meant "logical". The decision is now centralized and reasoned;
 * the cross-port byte-canonical + ADR-0023 (sealed agreed-vocabulary registry)
 * together guarantee that an `INCLUDED` attr really is agreed vocabulary rather
 * than an accidental registration, so inclusion-by-classification is sound.
 *
 * NOTE on liveness: a per-type attr exclusion (e.g. `extends`, `object`) is a
 * CROSS-PORT carve-out that only the OO ports (Java/Kotlin) physically register;
 * TS/C#/Python never register those names, so a single port's emitter cannot
 * judge a carve-out "dead" — that is a cross-port property, asserted by the
 * shared byte-canonical, not here.
 */
const INCLUDED = "included" as const;
export type AttrClassification = ExclusionReason | typeof INCLUDED;

export function classifyPerTypeAttr(name: string): AttrClassification {
  return EXCLUDED_PER_TYPE_ATTRS.get(name) ?? INCLUDED;
}

/**
 * Build one manifest type entry from its full TypeDefinition (FR-033). Key order
 * is fixed for byte-stability: `type`, `subType`, `description`, then optional
 * `rules`/`example`/`whenToUse` (emitted only when present), then `attrs`,
 * `children` (the sorted constraint graph), and optional `parents` (emitted only
 * when present + non-empty, sorted ASCII).
 */
function toManifestType(def: TypeDefinition): ManifestType {
  // Build in fixed key order. Optional docs facets (rules/example/whenToUse) sit
  // between `description` and `attrs`, so they are spread into the literal (when
  // present) to preserve insertion order — `attrs`/`children`/`parents` follow.
  const out: ManifestType = {
    type: def.typeId.type,
    subType: def.typeId.subType,
    description: def.description,
    ...(def.rules !== undefined ? { rules: def.rules } : {}),
    ...(def.example !== undefined ? { example: def.example } : {}),
    ...(def.whenToUse !== undefined ? { whenToUse: def.whenToUse } : {}),
    attrs: sortedPerTypeAttrs(def.attributes, def.typeId.type, def.typeId.subType),
    children: sortedChildren(def.childRules),
  };
  if (def.parents !== undefined && def.parents.length > 0) {
    out.parents = [...def.parents].sort(compareStrings);
  }
  return out;
}

/**
 * Build the canonical registry manifest object from an assembled registry.
 *
 * The registry must already be composed (e.g. `composeRegistry(coreProviders)`)
 * so all providers — core types, db-domain attrs, common doc attrs — have run.
 */
export function buildRegistryManifest(registry: TypeRegistry): RegistryManifest {
  // Walk every registered (type, subType). `allTypes()` returns the TypeIds;
  // `find` gives each one's full TypeDefinition (description / attributes /
  // childRules / parents / rules / example / whenToUse) in a single lookup.
  const types: ManifestType[] = registry
    .allTypes()
    // Skip excluded (type, subType) rows: the `metadata.base` inheritance
    // anchor (C-5) + the generic TS-presentation `view.*` controls (B-2).
    .filter((typeId) => !isExcludedTypeSubType(typeId.type, typeId.subType))
    .map((typeId) =>
      toManifestType(
        // The type IS registered (it came from allTypes()), so find() is defined.
        registry.find(typeId.type, typeId.subType) as TypeDefinition,
      ),
    )
    // Sort by the full "type.subType" key for a stable, port-independent order.
    .sort((a, b) =>
      compareStrings(`${a.type}.${a.subType}`, `${b.type}.${b.subType}`),
    );

  const commonAttrs = sortedAttrs(registry.getCommonAttrs());

  // defaultSubTypes: rebuild with sorted keys so JSON.stringify emits a stable
  // key order. There is no public "all default subTypes" accessor, so derive
  // the candidate type names from the registered types and probe each.
  const typeNames = Array.from(new Set(types.map((t) => t.type))).sort(
    compareStrings,
  );
  const defaultSubTypes: Record<string, string> = {};
  for (const typeName of typeNames) {
    const defaultSub = registry.defaultSubTypeOf(typeName);
    if (defaultSub !== undefined) {
      defaultSubTypes[typeName] = defaultSub;
    }
  }

  return { metamodelVersion: METAMODEL_VERSION, types, commonAttrs, defaultSubTypes };
}

/**
 * Emit the canonical registry manifest as a byte-stable JSON string.
 *
 * Serialization contract — every port MUST match this exactly:
 *  - 2-space indentation (JSON.stringify(_, _, 2)).
 *  - Object keys in a fixed order (JSON.stringify preserves insertion order):
 *    the manifest is built with `metamodelVersion` (first — the spec-version
 *    header), then `types`, `commonAttrs`, `defaultSubTypes`.
 *    - Each attr: `name`, `valueType`, `isArray`, `required`, then optional
 *      `allowedValues` (omitted unless the attr declares a non-empty closed set —
 *      decision 5), then `description`, then optional `rules`, `example`,
 *      `whenToUse` (each omitted when absent).
 *    - Each type: `type`, `subType`, `description`, then optional `rules`,
 *      `example`, `whenToUse` (omitted when absent), then `attrs`, `children`,
 *      then optional `parents` (omitted when absent/empty).
 *    - Each child (FR-033 constraint graph): `childType`, `childSubType`,
 *      `childName`, then optional `min`, `max`, `named` (each emitted ONLY when
 *      defined on the rule — legacy wildcard rules omit all three; `max: null`
 *      is emitted when the rule sets it null, omitted when undefined).
 *  - All arrays sorted: `types` by "type.subType"; each `attrs` by name;
 *    each type's `children` by the tuple (childType, childSubTypeKey, childName)
 *    where childSubTypeKey is the string or the comma-joined list; `parents`
 *    ascending; `commonAttrs` by name; `defaultSubTypes` keys sorted.
 *  - `valueType: null` literal for polymorphic/untyped attrs.
 *  - `childSubType` is a string OR a string[] (a list of admitted subtypes).
 *  - `description` is required + non-empty on every type and attr (gated by the
 *    coverage assertion in registry-conformance.test.ts — mirrors ADR-0023).
 *  - A single trailing newline (matches the repo's committed-canonical style).
 */
export function emitRegistryManifest(registry: TypeRegistry): string {
  const manifest = buildRegistryManifest(registry);
  return `${JSON.stringify(manifest, null, 2)}\n`;
}

File: server/python/src/metaobjects/registry_manifest.py
"""SP-G Registry Conformance — the Python registry-manifest emitter.

Walks an assembled core ``TypeRegistry`` and serializes the LOGICAL metamodel
vocabulary as a canonical, fully-sorted, byte-stable JSON manifest. This is the
single-source contract the other four ports (TS / C# / Java / Kotlin) must
byte-match — a structural gate against the SP-C class of silent vocabulary
drift (a port's registry diverging — wrong attr names, missing subtypes,
different required-ness — with every behavioral corpus still green).

The IN/OUT boundary (the v1 logical subset emittable byte-identically by all
five ports) is documented in ``fixtures/registry-conformance/README.md``. In
short: ``type.subType`` + ``attrs[{name, valueType, required, allowedValues?}]`` +
``commonAttrs`` + ``defaultSubTypes``. ``allowedValues`` (ADR-0036 Wave 1,
decision 5) is the closed value-set of a closed-enum attr — emitted ONLY when the
attr declares one, OMITTED for open / format-validated attrs (@currency, @locale).
EXCLUDED from v1 (per-port-physical or not-universally-tracked-on-the-registry):
factories/native bindings; ``AttrSchema.default``; ``inheritsFrom``;
``child_rules``.

The TS emitter (``server/typescript/packages/metadata/src/registry-manifest.ts``,
``emitRegistryManifest``) is the reference implementation; the canonical bytes
live in ``fixtures/registry-conformance/expected-registry.json``.
"""
from __future__ import annotations

import json
from enum import Enum

from .documentation.doc_constants import DOC_ATTR_DESCRIPTION
from .meta.core.attr.attr_constants import (
    ATTR_SUBTYPE_STRING,
    ATTR_SUBTYPE_STRINGARRAY,
)
from .meta.presentation.view.view_constants import VIEW_SUBTYPE_CURRENCY
from .registry import AttrSchema, ChildRule, TypeDefinition, TypeRegistry
from .shared.base_types import SUBTYPE_BASE, TYPE_ATTR, TYPE_METADATA, TYPE_VIEW
from .shared.structural import KEY_IS_ARRAY

# Wave 3b — the in/out boundary is an EXPLICIT CLASSIFICATION (a reason category
# per carve-out), not a bare name-match. The negative branch of a name-list
# silently meant "logical"; now ``classify_per_type_attr`` returns either an
# ``ExclusionReason`` (carved out, with a documented category) or ``INCLUDED``
# (logical cross-port vocabulary). Inclusion-by-classification is sound because
# ADR-0023 seals the agreed-vocabulary registry. The axis is
# cross-port-CONTRACT vs port-PRIVATE-mechanism (NOT abstract-vs-physical — the
# physical-DB attrs column/dbColumnType/db.indexed/precision/scale/maxLength/
# unique ARE logical here, the agreed persistence vocabulary). See
# fixtures/registry-conformance/README.md.


class ExclusionReason(str, Enum):
    """Reason a per-type attr/row is carved out of the agreed vocabulary."""

    #: Sentinel: NOT excluded — logical cross-port vocabulary.
    INCLUDED = "included"
    #: Native type-binding / factory (incl. ADR-0001 ``object``, ADR-0005 ``objectAdapter``).
    NATIVE_BINDING = "native-binding"
    #: Bare structural / OO-shape keyword (isArray/isAbstract/extends/implements/isInterface).
    STRUCTURAL_KEYWORD = "structural-keyword"
    #: A commonAttr (``description``) re-registered per-type — belongs in commonAttrs.
    COMMON_ATTR_DUP = "common-attr-dup"
    #: The ``metadata.base`` per-port inheritance anchor (deferred inheritsFrom facet).
    INHERITANCE_ANCHOR = "inheritance-anchor"
    #: TS-web-presentation-only facet (the generic ``view.*`` controls).
    PRESENTATION_ONLY = "presentation-only"
    #: FR-024 reference-first rollout slot (RETIRED at the atomic flip): the
    #: ``object.projection`` row + the ``origin.aggregate.via`` required-override
    #: were excluded until every port registered the FR-024 loader-grammar slice,
    #: then removed with the canonical updated in ONE commit (the ``@responseRef``
    #: carve-out-close playbook). Kept as the documented lifecycle slot for the
    #: next rollout; currently no members.
    FR024_PENDING = "fr024-pending"


_ATTR_NAME_IS_ABSTRACT = "isAbstract"
_ATTR_NAME_EXTENDS = "extends"
_ATTR_NAME_IMPLEMENTS = "implements"
_ATTR_NAME_IS_INTERFACE = "isInterface"
# ADR-0001 class-FQN type binding + ADR-0005 hybrid value-access seam.
_ATTR_NAME_OBJECT = "object"
_ATTR_NAME_OBJECT_ADAPTER = "objectAdapter"

# Per-type attr names carved out of the agreed vocabulary, each mapped to its
# PORT_PRIVATE reason. An attr NOT in this map is logical (INCLUDED) by the
# ADR-0023 sealed-vocabulary contract. ``description`` is carved out ONLY
# per-type — it stays in the commonAttrs block.
_EXCLUDED_PER_TYPE_ATTRS: dict[str, ExclusionReason] = {
    KEY_IS_ARRAY: ExclusionReason.STRUCTURAL_KEYWORD,
    _ATTR_NAME_IS_ABSTRACT: ExclusionReason.STRUCTURAL_KEYWORD,
    _ATTR_NAME_EXTENDS: ExclusionReason.STRUCTURAL_KEYWORD,
    _ATTR_NAME_IMPLEMENTS: ExclusionReason.STRUCTURAL_KEYWORD,
    _ATTR_NAME_IS_INTERFACE: ExclusionReason.STRUCTURAL_KEYWORD,
    _ATTR_NAME_OBJECT: ExclusionReason.NATIVE_BINDING,
    _ATTR_NAME_OBJECT_ADAPTER: ExclusionReason.NATIVE_BINDING,
    DOC_ATTR_DESCRIPTION: ExclusionReason.COMMON_ATTR_DUP,
}


def classify_per_type_attr(name: str) -> ExclusionReason:
    """Classify a per-type attr: an ``ExclusionReason`` (carved out) or
    ``ExclusionReason.INCLUDED`` (logical). Total — no silent default."""
    return _EXCLUDED_PER_TYPE_ATTRS.get(name, ExclusionReason.INCLUDED)


def classify_type_subtype(type_name: str, sub_type: str) -> ExclusionReason:
    """Classify a ``(type, subType)`` row: the metadata.base inheritance anchor
    (C-5) / the generic ``view.*`` presentation controls (B-2) / INCLUDED."""
    if type_name == TYPE_METADATA and sub_type == SUBTYPE_BASE:
        return ExclusionReason.INHERITANCE_ANCHOR  # C-5 — Java's internal inheritance anchor
    if type_name == TYPE_VIEW and sub_type not in (SUBTYPE_BASE, VIEW_SUBTYPE_CURRENCY):
        return ExclusionReason.PRESENTATION_ONLY  # B-2 — TS-web-presentation generic view controls
    return ExclusionReason.INCLUDED


def _is_excluded_type_subtype(type_name: str, sub_type: str) -> bool:
    """True if a ``(type, subType)`` row is carved out of the manifest (any reason)."""
    return classify_type_subtype(type_name, sub_type) is not ExclusionReason.INCLUDED


def _to_manifest_attr(attr: AttrSchema) -> dict[str, object]:
    """Normalize one AttrSchema to the manifest's logical attr shape.

    Emits ``{name, valueType, isArray, required, allowedValues?}`` — decomposing
    array-ness into a scalar ``valueType`` + an orthogonal ``isArray`` flag.
    ``allowedValues`` is emitted only when the attr declares a non-empty closed
    set (ADR-0036 Wave 1, decision 5); ``default`` is intentionally dropped.
    ``value_type`` is ``None`` for polymorphic/untyped attrs (e.g. ``@default``);
    the manifest renders that as an explicit JSON ``null``. A legacy
    ``stringarray`` value_type token is decomposed to
    ``{valueType: "string", isArray: true}`` so no ``stringarray`` token reaches
    the manifest.
    """
    is_legacy_string_array = attr.value_type == ATTR_SUBTYPE_STRINGARRAY
    is_array = attr.is_array or is_legacy_string_array
    value_type = ATTR_SUBTYPE_STRING if is_legacy_string_array else attr.value_type
    # Fixed key order: name, valueType, isArray, required, allowedValues?,
    # description, then the optional rules/example/whenToUse (FR-033) — emitted
    # ONLY when present so absent keys stay absent (dict insertion order →
    # byte-stable).
    out: dict[str, object] = {
        "name": attr.name,
        "valueType": value_type,
        "isArray": is_array,
        "required": attr.required,
    }
    # The closed value-set of a closed-enum attr (ADR-0036 Wave 1, decision 5).
    # Emitted ONLY when the attr declares a non-empty allowed_values set; OMITTED
    # for open / format-validated attrs (e.g. @currency ISO-4217, @locale BCP-47,
    # @cardinality). Byte-gates closed-enum vocabularies cross-port. Inserted
    # BEFORE description so the key order is name → valueType → isArray → required
    # → (allowedValues?) → description → (rules?/example?/whenToUse?). Each member
    # is stringified for a stable cross-port surface.
    if attr.allowed_values is not None and len(attr.allowed_values) > 0:
        out["allowedValues"] = [str(v) for v in attr.allowed_values]
    out["description"] = attr.description
    if attr.rules is not None:
        out["rules"] = attr.rules
    if attr.example is not None:
        out["example"] = attr.example
    if attr.when_to_use is not None:
        out["whenToUse"] = attr.when_to_use
    return out


def _sorted_attrs(attrs: list[AttrSchema]) -> list[dict[str, object]]:
    """Sort attrs by name (ascending, ASCII codepoint compare)."""
    return [_to_manifest_attr(a) for a in sorted(attrs, key=lambda a: a.name)]


# FR-024-pending manifest requiredness overrides (keyed ``type.subType.attr``) —
# the attr-level analogue of the FR024_PENDING row carve-out. EMPTY since the
# atomic all-ports flip (``origin.aggregate.via`` is optional everywhere,
# ADR-0029 d5). The mechanism stays for the next reference-first rollout.
_FR024_PENDING_REQUIRED_OVERRIDES: dict[str, bool] = {}


def _sorted_per_type_attrs(
    attrs: list[AttrSchema], type_name: str = "", sub_type: str = ""
) -> list[dict[str, object]]:
    """As ``_sorted_attrs``, but keeping only attrs the explicit classification
    marks ``INCLUDED`` (logical cross-port vocabulary). A carved-out attr
    (structural keyword, native binding, per-type ``description`` dup) is dropped
    for a documented reason, never a silent name-match. Applied ONLY to per-type
    attrs — ``description`` stays in the commonAttrs block. The FR-024 pending
    required-overrides are applied here (keyed ``type.subType.attr``)."""
    rows = _sorted_attrs(
        [a for a in attrs if classify_per_type_attr(a.name) is ExclusionReason.INCLUDED]
    )
    for row in rows:
        override = _FR024_PENDING_REQUIRED_OVERRIDES.get(
            f"{type_name}.{sub_type}.{row['name']}"
        )
        if override is not None:
            row["required"] = override
    return rows


def _child_sub_type_key(child_sub_type: str | list[str]) -> str:
    """Canonical sort/dedupe key for a child rule's subType: the string itself,
    or a comma-joined list (matches the TS ``childSubTypeKey``)."""
    if isinstance(child_sub_type, list):
        return ",".join(child_sub_type)
    return child_sub_type


def _to_manifest_child(rule: ChildRule) -> dict[str, object]:
    """Normalize one structural ``ChildRule`` to the manifest's child shape (FR-033).

    Fixed key order: ``childType``, ``childSubType`` (a string, ``"*"``, or a
    list of admitted subtypes), ``childName``, then the optional cardinality
    ``min``/``max``/``named`` — emitted ONLY when set on the rule. ``max`` may be
    the explicit JSON ``null`` literal when declared-unbounded (``max_is_null``);
    an absent ``max`` is omitted entirely. Mirrors the Java ``ManifestChild``.
    """
    out: dict[str, object] = {
        "childType": rule.child_type,
        "childSubType": rule.child_sub_type,
        "childName": rule.child_name,
    }
    if rule.min is not None:
        out["min"] = rule.min
    if rule.max is not None or rule.max_is_null:
        out["max"] = rule.max  # None → JSON null literal (declared-unbounded)
    if rule.named is not None:
        out["named"] = rule.named
    return out


def _sorted_children(definition: TypeDefinition) -> list[dict[str, object]]:
    """Build the FR-033 constraint graph for one (type, subType) from its
    STRUCTURAL child rules — every ``ChildRule`` whose ``child_type`` is NOT the
    attr type (the strict model has no any-attr wildcard; an ``attr`` child rule
    is dropped). Sorted + de-duped by ``(childType, childSubTypeKey, childName)``,
    matching the TS reference's ``sortedChildren``.

    NOTE: Python's current ``child_rules`` are broad / carry no cardinality, so
    the emitted content will be wrong until S-B2a sources the strict graph from
    spec/metamodel/*.json — that is the expected intermediate (VALUE-level) diff.
    """
    by_key: dict[str, dict[str, object]] = {}
    for rule in definition.child_rules:
        if rule.child_type == TYPE_ATTR:
            continue  # attr requirement is the attrs block; strict model has no attr wildcard
        child = _to_manifest_child(rule)
        key = f"{rule.child_type} {_child_sub_type_key(rule.child_sub_type)} {rule.child_name}"
        by_key.setdefault(key, child)
    children = list(by_key.values())
    children.sort(
        key=lambda c: (
            c["childType"],  # type: ignore[index]
            _child_sub_type_key(c["childSubType"]),  # type: ignore[arg-type]
            c["childName"],  # type: ignore[index]
        )
    )
    return children


METAMODEL_VERSION = "0.10"
"""Rolled-up spec-version for the cross-port registry manifest.

``"0"`` = pre-1.0 / unstable (semver major-0). Flips to ``"1.0"`` at the
1.0 vocabulary cut. Emitted as the first top-level key in the canonical
manifest so every port's byte-comparison gate sees it at position 0.
"""


def build_registry_manifest(registry: TypeRegistry) -> dict[str, object]:
    """Build the canonical registry-manifest object from an assembled registry.

    The registry must already be composed (e.g.
    ``compose_registry([core_provider, doc_provider])``) so all providers — core
    types, the DB-domain attrs Python keeps on its field defs, and the common
    doc attrs — have run.

    All collections are sorted explicitly (not relying on dict insertion order)
    so the serialization is byte-stable and port-independent.
    """
    types: list[dict[str, object]] = []
    # Iterate every registered (type, subType). Sorting is applied after the
    # walk, so dict iteration order is irrelevant.
    for definition in registry._defs.values():  # noqa: SLF001 (no public iterator)
        if _is_excluded_type_subtype(definition.type, definition.sub_type):
            continue  # metadata.base anchor (C-5) / generic view.* controls (B-2)
        # Fixed TS key order: type, subType, description, optional
        # rules/example/whenToUse (only when present), attrs, children, optional
        # parents (only when non-empty + sorted ASCII). Insertion order on the
        # dict → byte-stable; absent optional keys are omitted entirely.
        type_dict: dict[str, object] = {
            "type": definition.type,
            "subType": definition.sub_type,
            "description": definition.description,
        }
        if definition.rules is not None:
            type_dict["rules"] = definition.rules
        if definition.example is not None:
            type_dict["example"] = definition.example
        if definition.when_to_use is not None:
            type_dict["whenToUse"] = definition.when_to_use
        type_dict["attrs"] = _sorted_per_type_attrs(
            definition.attrs, definition.type, definition.sub_type
        )
        type_dict["children"] = _sorted_children(definition)
        if definition.parents:
            type_dict["parents"] = sorted(definition.parents)
        types.append(type_dict)
    types.sort(key=lambda t: f"{t['type']}.{t['subType']}")

    common_attrs = _sorted_attrs(registry.get_common_attrs())

    # defaultSubTypes: probe each registered type name; emit with sorted keys.
    type_names = sorted({t["type"] for t in types})  # type: ignore[misc]
    default_sub_types: dict[str, str] = {}
    for type_name in type_names:
        default_sub = registry.default_sub_type_of(type_name)  # type: ignore[arg-type]
        if default_sub is not None:
            default_sub_types[type_name] = default_sub

    # Fixed top-level key order: metamodelVersion (C4), types, commonAttrs, defaultSubTypes.
    return {
        "metamodelVersion": METAMODEL_VERSION,
        "types": types,
        "commonAttrs": common_attrs,
        "defaultSubTypes": default_sub_types,
    }


def emit_registry_manifest(registry: TypeRegistry) -> str:
    """Emit the canonical registry manifest as a byte-stable JSON string.

    Serialization contract — every port MUST match this exactly:
     - 2-space indentation.
     - Object keys in a fixed order (``types`` / ``commonAttrs`` /
       ``defaultSubTypes``; each type ``type`` / ``subType`` / ``attrs``; each
       attr ``name`` / ``valueType`` / ``isArray`` / ``required`` /
       optional ``allowedValues`` / ``description``).
     - All arrays sorted: ``types`` by ``"type.subType"``; each ``attrs`` by
       name; ``commonAttrs`` by name; ``defaultSubTypes`` keys sorted.
     - ``valueType: null`` literal for polymorphic/untyped attrs.
     - A single trailing newline.

    ``json.dumps(indent=2, separators=(",", ": "))`` reproduces JS
    ``JSON.stringify(obj, null, 2)`` byte-for-byte for ASCII content: no
    trailing whitespace on container lines, ``": "`` after keys, ``,`` line
    separators. ``ensure_ascii=False`` keeps any non-ASCII verbatim (there is
    none in the core vocabulary, but it matches the JS contract).
    """
    manifest = build_registry_manifest(registry)
    return (
        json.dumps(
            manifest,
            indent=2,
            ensure_ascii=False,
            separators=(",", ": "),
        )
        + "\n"
    )
